runelite-client/src/main/java/net/runelite/client/plugins/worldhopper/WorldHopperPlugin.java

```diff
@@ -0,0 +1,538 @@
+			{
+				try
+				{
+					int totalRequirement = Integer.parseInt(world.getActivity().substring(0, world.getActivity().indexOf(" ")));
+
+					if (totalLevel >= totalRequirement)
+					{
+						types.remove(WorldType.SKILL_TOTAL);
+					}
+				}
+				catch (NumberFormatException ex)
+				{
+					log.warn(null, ex);
+				}
+			}
+
+			// Break out if we've found a good world to hop to
+			if (currentWorldTypes.equals(types))
+			{
+				break;
+			}
+		}
+		while (world != currentWorld);
+
+		if (world == currentWorld)
+		{
+			String chatMessage = new ChatMessageBuilder()
+				.append(ChatColorType.NORMAL)
+				.append("Couldn't find a world to quick-hop to.")
+				.build();
+
+			chatMessageManager.queue(QueuedMessage.builder()
+				.type(ChatMessageType.GAME)
+				.runeLiteFormattedMessage(chatMessage)
+				.build());
+		}
+		else
+		{
+			hop(world.getId());
+		}
+	}
+
+	private void hop(int worldId)
+	{
+		// Don't try to hop if the world doesn't exist
+		World world = worldResult.findWorld(worldId);
+		if (world == null)
+		{
+			return;
+		}
+
+		if (client.getGameState() == GameState.LOGIN_SCREEN)
+		{
+			// on the login screen we can just change the world ourself
+			final net.runelite.api.World rsWorld = client.createWorld();
+			rsWorld.setActivity(world.getActivity());
+			rsWorld.setAddress(world.getAddress());
+			rsWorld.setId(world.getId());
+			rsWorld.setPlayerCount(world.getPlayers());
+			rsWorld.setLocation(world.getLocation());
+			rsWorld.setTypes(toWorldTypes(world.getTypes()));
+
+			client.changeWorld(rsWorld);
+			return;
+		}
+
+		String chatMessage = new ChatMessageBuilder()
+			.append(ChatColorType.NORMAL)
+			.append("Quick-hopping to World ")
+			.append(ChatColorType.HIGHLIGHT)
+			.append(world + "")
+			.append(ChatColorType.NORMAL)
+			.append("..")
+			.build();
+
+		chatMessageManager
+			.queue(QueuedMessage.builder()
+				.type(ChatMessageType.GAME)
+				.runeLiteFormattedMessage(chatMessage)
+				.build());
+
+		clientThread.invokeLater(() ->
+		{
+			if (client.getWidget(WidgetInfo.WORLD_SWITCHER_LIST) == null)
+			{
+				client.openWorldHopper();
+				return false;
+			}
+
+			client.hopToWorld(worldId);
+			return true;
+		});
+	}
+
+	private ChatPlayer getChatPlayerFromName(String name)
+	{
+		String cleanName = Text.removeTags(name);
+
+		Friend[] friends = client.getFriends();
+
+		if (friends != null)
+		{
+			for (Friend friend : friends)
+			{
+				if (friend != null && friend.getName().equals(cleanName))
+				{
+					return friend;
+				}
+			}
+		}
+
+		ClanMember[] clanMembers = client.getClanMembers();
+
+		if (clanMembers != null)
+		{
+			for (ClanMember clanMember : clanMembers)
+			{
+				if (clanMember != null && clanMember.getUsername().equals(cleanName))
+				{
+					return clanMember;
+				}
+			}
+		}
+
+		return null;
+	}
+
+	private static EnumSet<net.runelite.api.WorldType> toWorldTypes(final EnumSet<net.runelite.http.api.worlds.WorldType> apiTypes)
```

Comment:
This is duplicate of DefaultWorld code, make it not, and create common method for it.